executor/src/wasm_executor.rs

```diff
@@ -136,90 +148,80 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
+	},
+	ext_enumerated_trie_root(values_data: *const u8, values_len: u32, lens_data: *const u32, lens_len: u32, result: *mut u8) => {
+		let values = (0..lens_len)
+			.map(|i| this.memory.read_primitive(lens_data + i * 4))
+			.collect::<::std::result::Result<Vec<u32>, DummyUserError>>()?
+			.into_iter()
+			.scan(0u32, |acc, v| { let o = *acc; *acc += v; Some((o, v)) })
+			.map(|(offset, len)|
+				this.memory.get(values_data + offset, len as usize)
+					.map_err(|_| DummyUserError)
+			)
+			.collect::<::std::result::Result<Vec<_>, DummyUserError>>()?;
+		let r = ordered_trie_root(values.into_iter());
+		this.memory.set(result, &r[..]).map_err(|_| DummyUserError)?;
 	},
 	ext_chain_id() -> u64 => {
 		this.ext.chain_id()
 	},
 	ext_twox_128(data: *const u8, len: u32, out: *mut u8) => {
-		let maybe_value = if len == 0 {
-			Ok(vec![])
-		} else {
-			this.memory.get(data, len as usize)
-		};
-		let result = if let Ok(value) = maybe_value {
-			twox_128(&value)
+		let result = if len == 0 {
+			twox_128(&[0u8; 0])
 		} else {
-			[0; 16]
+			twox_128(&this.memory.get(data, len as usize).map_err(|_| DummyUserError)?)
 		};
-		let _ = this.memory.set(out, &result);
+		this.memory.set(out, &result).map_err(|_| DummyUserError)?;
 	},
 	ext_twox_256(data: *const u8, len: u32, out: *mut u8) => {
-		let maybe_value = if len == 0 {
-			Ok(vec![])
+		let result = if len == 0 {
+			twox_256(&[0u8; 0])
 		} else {
-			this.memory.get(data, len as usize)
+			twox_256(&this.memory.get(data, len as usize).map_err(|_| DummyUserError)?)
 		};
-		let result = if let Ok(value) = maybe_value {
-			twox_256(&value)
-		} else {
-			[0; 32]
-		};
-		let _ = this.memory.set(out, &result);
+		this.memory.set(out, &result).map_err(|_| DummyUserError)?;
 	},
 	ext_blake2_256(data: *const u8, len: u32, out: *mut u8) => {
-		let maybe_value = if len == 0 {
-			Ok(vec![])
-		} else {
-			this.memory.get(data, len as usize)
-		};
-		let result = if let Ok(value) = maybe_value {
-			blake2_256(&value)
+		let result = if len == 0 {
+			blake2_256(&[0u8; 0])
 		} else {
-			[0; 32]
+			blake2_256(&this.memory.get(data, len as usize).map_err(|_| DummyUserError)?)
 		};
-		let _ = this.memory.set(out, &result);
+		this.memory.set(out, &result).map_err(|_| DummyUserError)?;
 	},
 	ext_ed25519_verify(msg_data: *const u8, msg_len: u32, sig_data: *const u8, pubkey_data: *const u8) -> u32 => {
-		(||{
-			let mut sig = [0u8; 64];
-			if let Err(_) = this.memory.get_into(sig_data, &mut sig[..]) {
-				return 2;
-			};
-			let mut pubkey = [0u8; 32];
-			if let Err(_) = this.memory.get_into(pubkey_data, &mut pubkey[..]) {
-				return 3;
-			};
-
-			if let Ok(msg) = this.memory.get(msg_data, msg_len as usize) {
-				if ed25519::Signature::from(sig).verify(&msg, &ed25519::Public::from(pubkey)) {
-					0
-				} else {
-					5
-				}
-			} else {
-				4
-			}
-		})()
+		let mut sig = [0u8; 64];
+		this.memory.get_into(sig_data, &mut sig[..]).map_err(|_| DummyUserError)?;
```

Comment:
Previous version preserved error code here. Is it intentionally discarded now?

Reply:
yup - any attempt to read out of bounds memory is now a strict runtime error.